### PR TITLE
Create an `Argument` class and allow convertion of optdict into them

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

## Description

This is my first proposed step towards #5392.

We could build on this in a separate branch, as I intend not to make this interfere too much with any current code. That should keep merge conflicts low.

All of this is based on the documentation of `argparse` found here:
https://docs.python.org/3/library/argparse.html

The idea of this PR is to showcase how we can transform our current `optdicts` into a `Argument` class that we can then at some point start passing to the `ArgumentsManager`. I think that second part would be the next step, to make this in a MVP. After that we can start working on adding new checkers.
I chose the logging module as it is a fairly easy module without only two options.

Don't hold back! Let me know what you think of this. All criticism is valuable!

**Note:** Ideally these commits wouldn't need to be squashed in a final merge. That makes cherry-picking later on much easier (see `2.0` branch in `astroid`). So after final review I'll revisit the commit history.

### CODE REVIEW

**Comment by DanielNoord** [comment]
If anybody could help me figure out why this fails on Windows that would be much appreciated!

**Comment by jacobtylerwalls** [comment]
I think it's that [subroutines can't be pickled on Windows](https://stackoverflow.com/a/8805244/14818839). Any way we can avoid pickling `ArgumentParser`?

**Comment by DanielNoord** [comment]
I'll try and come up with something!

**Comment by DanielNoord** [comment]
I have fixed the issue with `argparse.ArgumentsParser` pickling. It is annoying as it means we'll need to instantiate an `ArgumentsParser` and extract any relevant information during the init of the `ArgumentsManager`. 
On the other hand, that does force us to think clearly about how we handle arguments parsing and might help us avoid any redundant calls 😄

**Comment by DanielNoord** [comment]
Converting this to draft as I found issues with the current implementation. In retrospect I think it doesn't make sense to break this into smaller PRs, as we might run into problems after earlier parts have been merged.

Therefore, I'll keep working on this until we can completely use `argparse` to parse the options for the `logging` module. That should be a good test that at least basic functionality is guaranteed.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Do we really need new classes ? Or could we `use argparse.Namespace` directly ? optparse is more complex so if we try to do a "small refactor" we'll keep the optparse complexity in our implementation even when using argparse.

One thing I had in mind for this refactor was separating parsing option / checking. Maybe  creating NamedTuple or DataClasses and gives them to the checker pre-parsed instead of parsing the option inside each checker (in order to not have breaking change we could parse the option if the object is not given ?).

**Comment by DanielNoord** [comment]
> Do we really need new classes ? Or could we `use argparse.Namespace` directly ? optparse is more complex so if we try to do a "small refactor" we'll keep the optparse complexity in our implementation even when using argparse.

I think it is better to create new classes as it allows us to more easily work on this. If we start to inject into the pre-existing `optparse` based classes we will create merge conflicts very easily.

> One thing I had in mind for this refactor was separating parsing option / checking. Maybe creating NamedTuple or DataClasses and gives them to the checker pre-parsed instead of parsing the option inside each checker (in order to not have breaking change we could parse the option if the object is not given ?).

I'm not sure what you mean here.

I think the flow would be 1) register options on parser object, 2) parse options with parser object, 3) set namespace object as config namespace of linter object.

**Comment by DanielNoord** [comment]
This might be more difficult than I thought. I had a pretty good idea of how to not require `argparse.ArgumentParser` as subclass of `PyLinter`, but we need some way to store the `ArgumentParser`. Otherwise we would need to register all options every time we want to change any of them.

The problem is that even `Run` gets pickled, so I don't know a good place to register an `ArgumentParser` and store it.. I'll keep thinking about this, but this is a very large limitation of `argparse` which I'm not sure we can overcome..

**Comment by Pierre-Sassoulas** [comment]
> 3) set namespace object as config namespace of linter object.
 
 I agree with 1) and 2) but I don't understand this part.
  
> we need some way to store the ArgumentParser. Otherwise we would need to register all options every time we want to change any of them.

Couldn't we store the result of the parsing, i.e. an ``argparse.Namespace`` ?

**Comment by DanielNoord** [comment]
>  I agree with 1) and 2) but I don't understand this part.

After the parsing is done (ideally in the `Run` init) the resulting namespace needs to be made accessible to the linter class. So it can be accessed there.


> Couldn't we store the result of the parsing, i.e. an ``argparse.Namespace`` ?

There is a difference between the Namespace and the storage of arguments. The namespace only stores the current value whereas the parser objects also stores stuff like default value and description. I think it's important that information is not lost after initial parsing of arguments is done, so we'll preferably need to save both the parser and the namespace. Currently I don't see a way to store the parser, as it seems everything within Run is pickled..

**Comment by Pierre-Sassoulas** [comment]
>  the resulting namespace needs to be made accessible to the linter class. 

Ok I agree with that too.

> I think it's important that information is not lost after initial parsing of arguments is done,

Why ? Where could we use the default value and the description after the parsing is done ?

**Comment by DanielNoord** [comment]
> Why ? Where could we use the default value and the description after the parsing is done ?

We would need it to "reparse" for sub directories and for help messages. We would then need to recollect and reinstantiate all options for every subdirectory. That's not really good performance wise..,

**Comment by Pierre-Sassoulas** [comment]
Can't we parse all the options for each directories first and store them ? (By the way the configuration by directory feature was removed from flake8 because it was bringing "90% of the bugs"  we might want to reconsider that one)

**Comment by DanielNoord** [comment]
> Can't we parse all the options for each directories first and store them ? (By the way the configuration by directory feature was removed from flake8 because it was bringing "90% of the bugs" we might want to reconsider that one)

That might be possible, but that would definitely come later.

However, I think I might have found a way to do it. We won't store information about the arguments this way, but I think we could design everything in such a way that that is not necessary.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Would you consider using [`dill`](https://github.com/uqfoundation/dill) to pickle the `PyLinter` object for the multiprocessing handling? I'm basing this on this SO question which seems to suggest it should be able to fix our problem:
https://stackoverflow.com/questions/8804830/python-multiprocessing-picklingerror-cant-pickle-type-function

According to their README it has been download 200M+ times via Pypi and supports almost all python versions (`Python >=2.7, !=3.0.*`). Being able to pickle the input to `multiprocessing.Pool` would save all of the issues and allow us to make a more intuitive system.

---------------

Some additional arguments for why I'm exploring fixing the `multiprocessing` issue instead of working around: we can't be sure that all plugins and tools (such as prospector) create a `Run` class. In our own tests we don't do so, but after some additional thoughts I don't think we should do stuff in the `Run` `__init__` that is needed for `PyLinter` to work. (That's counterintuitive and bad design).
Thus, my third (fourth, fifth?) attempt at this doesn't really work. I hoped that by doing the argument parsing in `Run` we could skip the pickling.

**Comment by Pierre-Sassoulas** [comment]
Look like ``dill`` do not have any dependencies of its own and it's 3-clause BSD license seems compatible with pylint. We would more than double their download per day by adding them as a dependencies but it's already used a lot. 

> I don't think we should do stuff in the Run __init__ that is needed for PyLinter to work. (That's counterintuitive and bad design).

Depends on what we want in PyLinter, right now it's doing almost everything. There's a problem with breaking change in downstream plugins to consider for sure, but let's put that aside for a minute If I understood what we're talking about and the current state of the code, PyLinter (each checker even) is parsing the options right now ? Parsing the options first **before** launching anything and especially the parallel run actually make more sense imo ? i.e. PyLinter could be the class that analyses a set of file given a set of parsed options and have less responsibilities. Alternatively we could include multiprocessing handling in PyLinter if we want PyLinter and not Run to be the entrypoint ?

**Comment by DanielNoord** [comment]
> Look like `dill` do not have any dependencies of its own and it's 3-clause BSD license seems compatible with pylint. We would more than double their download per day by adding them as a dependencies but it's already used a lot.

I'll start work on a PR implementing it for our `parallel` run then, so we can continue discussion about actually doing this in a separate PR 😄 

> Depends on what we want in PyLinter, right now it's doing almost everything. There's a problem with breaking change in downstream plugins to consider for sure, but let's put that aside for a minute If I understood what we're talking about and the current state of the code, PyLinter (each checker even) is parsing the options right now ?

Since every `checker` is an `OptionsProvider` each plugin or checker added after initialising `PyLinter` can add new options that need to be read, "added to the parser" and then "parsed into the Namespace". The flow is as follows:
1. Initialise `Run`, 2. Initialise `PyLinter` and its associated `parser`, 3. "Add" all options from the `PyLinter` class, 4. "Parse" `PyLinter` options, 5. "Add" all options from default checkers and plugins, 6. "Parse" all options from default checkers and plugins, 7. "Add" all options from extra plugins, 8. "Parse" all options from extra plugins.

Note `load_defaults` in `register_checker` here:
https://github.com/PyCQA/pylint/blob/e815843293adf100662fe4e183c34f3040529a15/pylint/lint/pylinter.py#L754-L763

The problem is thus that at step 5 and step 7 we need to re-initialise an `ArgumentsParser` if we can't store the parser as an attribute or subclass of `PyLinter`. Adding the parser to `Run` doesn't help as for step 7 we don't have access to that step anymore.
I have tested what would happen if we had to re-initialise an `ArgumentsParser` 100 times and it became a significant impact on runtime.

> Parsing the options first **before** launching anything and especially the parallel run actually make more sense imo ? i.e. PyLinter could be the class that analyses a set of file given a set of parsed options and have less responsibilities. Alternatively we could include multiprocessing handling in PyLinter if we want PyLinter and not Run to be the entrypoint ?

We can't really do this as we need to parse the `load-plugins` option before knowing all the options we have to parse.

**Comment by Pierre-Sassoulas** [comment]
> We can't really do this as we need to parse the load-plugins option before knowing all the options we have to parse.

I'm probably off base here but what if we first parse ``load-plugin`` alone, then add their options in 5 at the same time than default checker and plugins ? And launch multiprocessing only after that. Because right now multiprocessing happens in 2., right ? Does'nt it happens too soon and it would be more efficient to do it later when we know exactly what "work" we'll have to do ?

**Comment by DanielNoord** [comment]
> I'm probably off base here but what if we first parse `load-plugin` alone, then add their options in 5 at the same time than default checker and plugins ?

We already kind of do this, but only in `Run`. If we want to have the `PyLinter` class work on its own (like it needs to do in `prospector`) we will need to the argument parsing in `PyLinter` and thus create an `ArgumentParser` there. If we want to allow plugins to add new options (which we should) we will still need to store (or re-initialize) a parser each time we register one.

I think I looked into loading all plugins, "adding" all options and then "parsing" them but stopped exploring it further as it would mean a complete rewrite of the `__init__` of `PyLinter`. I can look into it again though.

> And launch multiprocessing only after that. Because right now multiprocessing happens in 2., right ? Does'nt it happens too soon and it would be more efficient to do it later when we know exactly what "work" we'll have to do ?

I believe multiprocessing happens after step 6 now.

Another hack that is the result of the need to pickling (I believe) is that we set `linter.reporter` to `None` here:
https://github.com/PyCQA/pylint/blob/532be9afa75bd090a389db0f172c47ba9d37b411/pylint/lint/parallel.py#L106-L113
See this commit: https://github.com/Pierre-Sassoulas/pylint/commit/8babeffde54a696e7ba84d69f8fc1f871e432950


I'm going to see if I can "fix" the multiprocessing first, as it would allow for a much cleaner design and worry about the order of argument registering and parsing after. If that's okay with you of course!

**Comment by Pierre-Sassoulas** [comment]
> I'm going to see if I can "fix" the multiprocessing first, as it would allow for a much cleaner design and worry about the order of argument registering and parsing after

Sounds good, the order of arguments seems to be optimal right now anyway, thank you for clarifying

**Comment by DanielNoord** [comment]
Another iteration of this... But this has passed CI on my own fork already, so should be more robust.
Some points:

1. This is obviously blocked by #5609 
2. Originally I intended to work on this in a separate branch. However, after looking at the performance impact of the current version I think this might not be necessary. We spend less than 15 ms in the `arguments_manager` file per run, where 7 ms comes from `_parse_configuration_file` which should not take too much extra time as we add more checkers to the new parsing method (its runtime is not currently linearly dependent on the amount of checkers). Besides, at some point we could also looking into not registering "argparse-using" checkers in the optparse manager, which would reduce performance impact of migration as well.
3. The way the migration works is that each checker can pass `future_option_parsing=True` to the `__init__` of `BaseChecker` to start parsing with `argparse`. As such we can do the change checker per checker, tackling any issues as we go along. For example, in the current version not all `optdict` keys are supported and they can be added when we get to them. There are some additional TODO's that should be fixed as well, but aren't necessary for the base functionality which is shown here.
4. In the current version we don't need a `OptionsProviderMixIn` as everything is handled within 6 lines in `BaseChecker` and the actual `ArgumentsManager`. I hope this can stay true as it would remove complexity from the code.
5. I have opted to rewrite all validators for options. This gives me (and others) some extra work, but make sure we can completely remove the `optparse` files when we're done and the deprecation-period is over and allows me to correctly type everthing. One of the issues with the current system is that we have no good way to know what type everything can be, this can be avoided by rewriting everything from the ground up.

Please don't hold back on any criticism. I think this is the right way forward, but we only get one opportunity to fix this so we should be really considerate of any potential problems and issues.

File: tests/config/data/logging_format_interpolation_style.py
Comment (DanielNoord) [inline tests/config/data/logging_format_interpolation_style.py:1]: @Pierre-Sassoulas Can we exclude the copyright hook for this directory?

Comment (Pierre-Sassoulas) [reply tests/config/data/logging_format_interpolation_style.py:1]: Sure !

File: .pre-commit-config.yaml
Comment (Pierre-Sassoulas) [inline .pre-commit-config.yaml:72]: That's great, maybe we could add it line 16 in the "fixtures" variable too ?

File: pylint/config/utils.py
Comment (Pierre-Sassoulas) [inline pylint/config/utils.py:24]: Can this be seen by end user with custom plugin ? Maybe we could add a link to our issue tracker ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Look good already. I only skimmed because I'm not an expert in argparse custom parser and even less so in optparse so it's a hard review for me.

File: pylint/config/utils.py
Comment (DanielNoord) [reply pylint/config/utils.py:24]: I don't think so (right now). Plugins will need to start using the `argparse` system and as long as we don't tell about them they probably won't.

I hope to make `argparse` the standard for `pylint's` own checkers in `2.14` and then start telling people about them. That would be after we fix all these open issues though.

File: .pre-commit-config.yaml
Comment (DanielNoord) [reply .pre-commit-config.yaml:72]: I'll do a separate PR for the pre-commit changers

Comment (DanielNoord) [reply .pre-commit-config.yaml:72]: See #5974.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Do you want others to review this as well before merging? @jacobtylerwalls possibly?

**Comment by Pierre-Sassoulas** [comment]
More reviews is always a good thing if we can afford the luxury :)

File: pylint/config/config_initialization.py
Comment (jacobtylerwalls) [inline pylint/config/config_initialization.py:72]: Why is this still needed (given that it was moved below)?

File: pylintrc
Comment (jacobtylerwalls) [inline pylintrc:396]: When I revert this change I don't see a problem (locally). Problem on CI?

File: tests/config/test_argparse_config.py
Comment (jacobtylerwalls) [inline tests/config/test_argparse_config.py:32]:
```suggestion
    """Tests for the argparse implementation of OptionsProviderMixIn.
```

File: tests/config/test_argparse_config.py
Comment (jacobtylerwalls) [inline tests/config/test_argparse_config.py:39]:
```suggestion
        """Check that we raise messages when we do not supply any options."""
```

File: pylint/config/argument.py
Comment (jacobtylerwalls) [inline pylint/config/argument.py:41]: Maybe `flags`? or `names`? It's odd for `name` to be a list. (And most often a list of one name...)

File: pylint/config/arguments_manager.py
Comment (jacobtylerwalls) [inline pylint/config/arguments_manager.py:33]:
```suggestion
        """Register an options provider and load its defaults."""
```

File: pylint/config/arguments_manager.py
Comment (jacobtylerwalls) [inline pylint/config/arguments_manager.py:74]: May I ask if there is a reason we can't just permit other actions like `store_true` and send them along? As in, is this something a plugin writer could take advantage of? (I'm used to `store_true` and the docs/tutorial suggest it also...)

File: pylint/config/arguments_manager.py
Comment (jacobtylerwalls) [inline pylint/config/arguments_manager.py:87]: This is probably a nothing. But, since you know the parsing flow very well, I'll ask -- is there a chance `opt` is already in `arguments`? And if so, does `argparse` pick the right one later?

**Review by jacobtylerwalls — COMMENTED** [review]
Terrific, thanks for moving this migration forward. A few questions, some for my own understanding.

File: pylint/config/config_initialization.py
Comment (DanielNoord) [reply pylint/config/config_initialization.py:72]: I think you're referring to `linter._parse_plugin_configuration()`?

That has not been implemented yet. The code below is just the order in which things should occur. This is "captured" by `# TODO: Start of the implemenation of argument parsing with argparse` and the `TODO` in the method definition.

File: pylintrc
Comment (DanielNoord) [reply pylintrc:396]: I think this might become problematic at some point. Perhaps it was in an earlier version but doing `namespace.setting` will raise a `no-member` for `setting` on `namespace`. I could remove this for now, but we'll likely need it at some point.

File: pylint/config/argument.py
Comment (DanielNoord) [reply pylint/config/argument.py:41]: The documentation refers to it as `name or flags`. I will change it to flags.

File: pylint/config/arguments_manager.py
Comment (DanielNoord) [reply pylint/config/arguments_manager.py:74]: We should at some point. For now we only need `store` as logging only has `store` options (that's why I chose to implement there first).

The `_Argument` class also does not support of parameters that can be passed to `add_argument()`. I want to add these on a if-needed basis when I convert more checkers.

File: pylint/config/arguments_manager.py
Comment (DanielNoord) [reply pylint/config/arguments_manager.py:87]: Hm, I don't actually know how double arguments are picked up by the `config_parser`. Although I think that it will probably raise a `configparser.DoubleKeyError` or something. (`configparser` is not our own code).

The arguments in the namespace get overwritten if they get re-parsed. That's why the order in `config_initialization` is important.

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
Thanks for answering my questions. Is now a good time to edit the branch history?

**Comment by DanielNoord** [comment]
I guess so. I'll do a rebase.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Please merge if you're okay with this and the number of reviews 😄

**Review by Pierre-Sassoulas — APPROVED** [review]
Great refactor @DanielNoord !

**Pull request merged**
